### PR TITLE
ARTEMIS-1499 ProcessorBase warns if a flush is issued within the event loop

### PR DESCRIPTION
**Pull request by franz1981**
It provides the above warn and an improvement of task submissions within the event loop.

### CODE REVIEW

**Comment by franz1981** [comment]
@gemmellr @clebertsuconic I'm not very used to "even loop" patterns hence I call for help :)

I've found that our `ArtemisExecutor` deadlock with the (fresh new) test `shouldSubmitBlockingTasksFromEventLoopMaintainingOrder` TL;DR if a `CountDownLatch::await` is blocking the event loop while waiting a submitted `CoundDownLatch::countDown`.
Enforcing any code rule to avoid these patterns to be used it not simple, hence I've provided a fix to it.
Just as a note, the same pattern (verified) blocks the Netty's `EventLoop` too.
This fix come at the cost of a potential bigger calls stack, but providing a good boost for "in event loop" executions too, but I'm not 100% to have messed up with others things.
Wdyt?

**Comment by franz1981** [comment]
@clebertsuconic Re `ArtemisExecutor::flush` I've provided the most defensive/conservative way to manage it, but I'm sure that it can be solved (like `execute`) in a more permissive and simpler way if called inside the `event loop`.

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/ProcessorBase.java [hunk 1]
```diff
@@ -109,12 +125,49 @@ public final boolean isFlushed() {
       return stateUpdater.get(this) == STATE_NOT_RUNNING;
    }
 
-   protected void task(T command) {
```
Comment (clebertsuconic) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/ProcessorBase.java]: What about just fix the offending flush than add this complex piece?

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/ProcessorBase.java [hunk 2]
```diff
@@ -109,12 +125,49 @@ public final boolean isFlushed() {
...
-      tasks.add(command);
```
Comment (clebertsuconic) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/ProcessorBase.java]: -1.... lets just fix the offending part?

Comment (franz1981) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/ProcessorBase.java]: It is not only the flush that has issues but the execute too: take a look at the tests I've added :)

**Comment by clebertsuconic** [comment]
throwing an executor call, and waiting for a .wait is clearly a bug... 

Adding this complex code on such a hot path has more risks in itself than eventually fixing issues.

I wouldn't merge this PR

**Comment by clebertsuconic** [comment]
@franz1981  shouldSubmitBlockingTasksFromEventLoopMaintainingOrder is a buggy code... it's an anti-pattern we shouldn't have in our code. 

Sometimes we must do such thing, but then we must make sure we are using proper executions.. or proper callbacks (even better)

**Comment by franz1981** [comment]
i understand it and I agree, but please reconsider it with tests and a longer review: as I've said we can't prohibit that kind of behaviours with ease and if it works (as I've tested/expected) it could make everything more stable/solid.

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/ProcessorBase.java (hunk 2 above)
Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/ProcessorBase.java]: you've added tests that are simulating bugs at the code.. I would rather fix bugs than protect the executor from them.

**Comment by clebertsuconic** [comment]
@franz1981 I disagree.. it would hide bugs.. which could end up in being more dangerous.

**Comment by andytaylor** [comment]
Clebert is correct, lets fix rather than re engineer

**Comment by franz1981** [comment]
@andytaylor @clebertsuconic Got it :+1: 
I will refactor the current PR handling only the fix on `flush` while leaving the improvement for recursive executions to be proposed in a different PR :)

**Comment by franz1981** [comment]
@clebertsuconic Just a couple questions (so I can add some doc/tests): 

1. it is allowed/legal to submit new tasks` from the event loop thread?
2. it is allowed call `flush` from the event loop thread? What is the expected behaviour while doing it?

**Comment by clebertsuconic** [comment]
@franz1981 what fix on flush? where do you see it being used.


1. it is legal to submit tasks...
2. you shouldn't call flush within a thread runnable. That's broken behaviour.

Instead of fixing what's not broken... just remove any calls to flush within executors.. I don't think there's any now.

**Comment by clebertsuconic** [comment]
@franz1981 changing flush is still re-engineering..

if there's any code calling flush within the OrderedExecutor, that's a classical starvation problem.. it's a bug that needs to be removed.

Instead of changing the semantic of flush, lets just avoid the bug... this PR should be closed IMO.

**Comment by franz1981** [comment]
@clebertsuconic 
> Instead of changing the semantic of flush, lets just avoid the bug... this PR should be closed IMO.

I'm not sure about it TBH: help to understand why we can't check if `flush` is being called for within the event loop? 
It is a wrong use of it as you said and we can check/validate it with ease (or at least we could log it as a warning/error) and without any perf impact.

**Comment by clebertsuconic** [comment]
@franz1981 ProcessorBase has no knowledge of Netty. just make the fix simple.. flush is only called in 3 places.. fix those places instead of rewriting the whole thing.

**Comment by franz1981** [comment]
@clebertsuconic I've mentioned Netty just as an example of a event loop framework that doesn't handle well recursive submits into the event loop/blocking call issued from without the event loop.
The fix (if i'll remove the other changes) is simple: i'm just checking if the call will happen inside the event loop: let me push it and you can see it by yourself :+1:

**Comment by clebertsuconic** [comment]
I don't think we should have any thread locals on these executors... they are quite hot on the path.. I would leave it alone.. if there's anything broken.. just change the caller.

**Comment by franz1981** [comment]
@clebertsuconic Why a thread local is not ok? It is not allocating nothing and it costs about like a MOV assembly instruction as it is implemented on the last JDK: having such logs (in debug mode at least)  could be really of some help for a developer if some weird execution paths could happen...

**Comment by clebertsuconic** [comment]
@franz1981  is there actually an issue? an actual deadlock?

My problem is adding complexity to the code for no reason.

**Comment by franz1981** [comment]
@clebertsuconic You're right, but I will run some tests with the broker to see if the warning will be printed...just in case.
I've added a lil improvement on task submission but for the reasons you've said I can remove it and let it being pushed as an improvement in another PR (in the future).

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/ProcessorBase.java [hunk 3]
```diff
@@ -111,7 +141,10 @@ public final boolean isFlushed() {
 
    protected void task(T command) {
       tasks.add(command);
-      startPoller();
+      //there is no need to alert the poller if already in the event loop
```
Comment (franz1981) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/ProcessorBase.java]: This is the improvement I've mentioned: I've left it here just to understand if it is safe but I can push in another PR apart (in the future) too, if it seems safe.

Comment (franz1981) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/ProcessorBase.java]: @clebertsuconic This is the improvement I've mentioned: I've left it here just to understand if it is safe but I can push in another PR apart (in the future) too, if it seems safe.

**Comment by franz1981** [comment]
@clebertsuconic after several tests I haven't found any call to 'flush' from within the event loop thread: I can close this if isn't needed :)
Thanks for the reviews guys

**Comment by clebertsuconic** [comment]
@franz1981 yeah.. I think we should close this... because:

It's not actually even loop. A ProcessorBase is just the context of a fake executor. Each OrderedExecutor will have its own instance.. This thing may confuse things a bit.

Say you have 10 running sessions doing deliveries. .you will end up with 10 threads.. each an instance of the OrderedExecutor.. the only difference is that this will take a thread from the pool and it may schedule tasks on that.

**Pull request closed without merging**